client/api/hijack.go

```diff
@@ -1,24 +1,64 @@
-package client
+package api
 
 import (
 	"fmt"
 	"io"
 	"net"
 	"net/http"
 	"net/http/httputil"
+	"net/url"
 	"strings"
 	"time"
 
 	"github.com/hyperhq/hyper/lib/promise"
 	"github.com/hyperhq/hyper/utils"
-	"github.com/hyperhq/runv/lib/term"
 )
 
-func (cli *HyperClient) dial() (net.Conn, error) {
+func (cli *Client) hijackRequest(method, tag string, v *url.Values, tty bool, stdin io.ReadCloser, stdout, stderr io.Writer) error {
+	var (
+		hijacked = make(chan io.Closer)
+		errCh    chan error
+	)
+	// Block the return until the chan gets closed
+	defer func() {
+		if _, ok := <-hijacked; ok {
+			fmt.Printf("Hijack did not finish (chan still open)\n")
+		}
+	}()
+
+	request := fmt.Sprintf("/%s?%s", method, v.Encode())
+
+	errCh = promise.Go(func() error {
+		return cli.hijack("POST", request, tty, stdin, stdout, stderr, hijacked, nil, "")
+	})
+
+	// Acknowledge the hijack before starting
+	select {
+	case closer := <-hijacked:
+	// Make sure that hijack gets closed when returning. (result
+	// in closing hijack chan and freeing server's goroutines.
+		if closer != nil {
+			defer closer.Close()
+		}
+	case err := <-errCh:
+		if err != nil {
+			fmt.Printf("Error hijack: %s", err.Error())
+			return err
+		}
+	}
+
+	if err := <-errCh; err != nil {
+		fmt.Printf("Error hijack: %s", err.Error())
+		return err
+	}
+	return nil
+}
+
+func (cli *Client) dial() (net.Conn, error) {
 	return net.Dial(cli.proto, cli.addr)
 }
 
-func (cli *HyperClient) hijack(method, path string, setRawTerminal bool, in io.ReadCloser, stdout, stderr io.Writer, started chan io.Closer, data interface{}, hostname string) error {
+func (cli *Client) hijack(method, path string, setRawTerminal bool, in io.ReadCloser, stdout, stderr io.Writer, started chan io.Closer, data interface{}, hostname string) error {
 	defer func() {
```

Comment:
setRawTerminal  is not a part of the api. it needs to be handle outside of the hijack()